### PR TITLE
Screwdriver Pen

### PR DESCRIPTION
**Pull request by etherware-novice**
## About The Pull Request
Screwdriver pen, you can toggle it out to use it like a screwdriver! It also wont fit into your pda when folded out, so you have to

## Why It's Good For The Game
Port of https://github.com/BeeStation/BeeStation-Hornet/pull/7366

Its a fun little maint spawn that saves an item slot, while still being outmatched by more "official" solutions (hand drill, implant, just getting a toolbelt, etc)

## Changelog

:cl:
add: screwdriver pen
/:cl:

### CODE REVIEW

File: code/modules/paperwork/pen.dm [hunk 1]
```diff
@@ -339,3 +339,54 @@
...
+	else
+		extended = TRUE
+		tool_behaviour = TOOL_SCREWDRIVER
+		w_class = WEIGHT_CLASS_SMALL  // still can fit in pocket
+		force = 4  // copies force from screwdriver
+		throwforce = 5
+		throw_speed = 3
+		throw_range = 5
+		to_chat(user, span_notice("You extend the screwdriver."))
+	update_icon()
```
Comment (ShizCalev) [inline code/modules/paperwork/pen.dm:371]: update_appearance(UPDATE_ICON)

File: code/modules/paperwork/pen.dm [hunk 2]
```diff
@@ -339,3 +339,54 @@
 	force = 5
 	wound_bonus = 100
 	demolition_mod = 9000
+
+// screwdriver pen!
+
+/obj/item/pen/screwdriver
+	var/extended = FALSE
```
Comment (ShizCalev) [inline code/modules/paperwork/pen.dm:346]: new vars on the bottom.

File: code/modules/paperwork/pen.dm [hunk 3]
```diff
@@ -339,3 +339,54 @@
...
+
+	else
+		extended = TRUE
+		tool_behaviour = TOOL_SCREWDRIVER
+		w_class = WEIGHT_CLASS_SMALL  // still can fit in pocket
+		force = 4  // copies force from screwdriver
+		throwforce = 5
+		throw_speed = 3
+		throw_range = 5
+		to_chat(user, span_notice("You extend the screwdriver."))
```
Comment (ShizCalev) [inline code/modules/paperwork/pen.dm:370]: AddElement(/datum/element/eyestab)

File: code/modules/paperwork/pen.dm [hunk 4]
```diff
@@ -339,3 +339,54 @@
...
+
+/obj/item/pen/screwdriver/attack_self(mob/living/user)
+	if(extended)
+		extended = FALSE
+		w_class = initial(w_class)
+		tool_behaviour = initial(tool_behaviour)
+		force = initial(force)
+		throwforce = initial(throwforce)
+		throw_speed = initial(throw_speed)
+		throw_range = initial(throw_range)
```
Comment (ShizCalev) [inline code/modules/paperwork/pen.dm:359]: RemoveElement(/datum/element/eyestab)

File: code/modules/paperwork/pen.dm [hunk 5]
```diff
@@ -339,3 +339,54 @@
...
+		return ..()
+	if(HAS_TRAIT(user, TRAIT_PACIFISM))
+		to_chat(user, span_warning("You don't want to harm [M]!</span>"))
+		return
+	if(HAS_TRAIT(user, TRAIT_CLUMSY) && prob(50))
+		M = user
+	return eyestab(M,user)
+
+/obj/item/pen/screwdriver/update_icon()
+	var/base = initial(icon_state)
```
Comment (ShizCalev) [inline code/modules/paperwork/pen.dm:381]: update_icon_state()
. = ..()
icon_state = "[initial(icon_state)][extended ? "out":null]

File: code/modules/paperwork/pen.dm [hunk 6]
```diff
@@ -339,3 +339,47 @@
...
+/obj/item/pen/screwdriver
+	/// whether the pen is extended
+	var/extended = FALSE
+	desc = "A pen with an extendable screwdriver tip. This one has a yellow cap."
+	icon_state = "pendriver"
+	toolspeed = 1.20  // gotta have some downside
+
+/obj/item/pen/screwdriver/Initialize(mapload)
+	. = ..()
+	AddElement(/datum/element/eyestab)
```
Comment (ShizCalev) [inline code/modules/paperwork/pen.dm:354]: not in init, this should be in attack_self.

File: code/modules/paperwork/pen.dm [hunk 7]
```diff
@@ -339,3 +339,40 @@
...
+		extended = TRUE
+		tool_behaviour = TOOL_SCREWDRIVER
+		w_class = WEIGHT_CLASS_SMALL  // still can fit in pocket
+		force = 4  // copies force from screwdriver
+		throwforce = 5
+		throw_speed = 3
+		throw_range = 5
+		AddElement(/datum/element/eyestab)
+		to_chat(user, span_notice("You extend the screwdriver."))
+	update_apperance(UPDATE_ICON)
```
Comment (ShizCalev) [inline code/modules/paperwork/pen.dm:374]:
```suggestion

/obj/item/pen/screwdriver/vv_edit_var(var_name, var_value)
	if(var_name == NAMEOF(src, extended))
		if(var_value != extended)
			toggle_screwdriver()
			datum_flags |= DF_VAR_EDITED
			return
	return ..()

/obj/item/pen/screwdriver/proc/toggle_screwdriver(mob/user)
	extended = !extended
	if(user)
		balloon_alert(user, "[extended ? "extended" : "retracted"]!")
		
	if(extended)
		w_class = initial(w_class)
		tool_behaviour = initial(tool_behaviour)
		force = initial(force)
		throwforce = initial(throwforce)
		throw_speed = initial(throw_speed)
		throw_range = initial(throw_range)
		RemoveElement(/datum/element/eyestab)
	else
		tool_behaviour = TOOL_SCREWDRIVER
		w_class = WEIGHT_CLASS_SMALL  // still can fit in pocket
		force = 4  // copies force from screwdriver
		throwforce = 5
		throw_speed = 3
		throw_range = 5
		AddElement(/datum/element/eyestab)
	update_apperance(UPDATE_ICON)
	
/obj/item/pen/screwdriver/attack_self(mob/living/user)
	toggle_screwdriver(user)

```

Comment (ShizCalev) [reply code/modules/paperwork/pen.dm:374]: varedit support. ♥

File: code/modules/paperwork/pen.dm [hunk 8]
```diff
@@ -339,3 +339,52 @@
...
+			datum_flags |= DF_VAR_EDITED
+			return
+	return ..()
+
+/obj/item/pen/screwdriver/proc/toggle_screwdriver(mob/user)
+	extended = !extended
+	if(user)
+		balloon_alert(user, "[extended ? "extended" : "retracted"]!")
+		
+	if(extended)
```
Comment (ShizCalev) [inline code/modules/paperwork/pen.dm:366]:
```suggestion
	if(!extended)
```

File: code/_globalvars/lists/maintenance_loot.dm [hunk 9]
```diff
@@ -84,6 +84,7 @@ GLOBAL_LIST_INIT(common_loot, list( //common: basic items
 		/obj/item/weldingtool = 1,
 		/obj/item/wirecutters = 1,
 		/obj/item/wrench = 1,
+		/obj/item/pen/screwdriver = 1
```
Comment (optimumtact) [inline code/_globalvars/lists/maintenance_loot.dm:87]: should not be in the common loot list

Comment (optimumtact) [reply code/_globalvars/lists/maintenance_loot.dm:87]: uncommon would be fine

**Comment by optimumtact** [comment]
really gonna do me like that 
picture of oranges assigning good first PR label before contributor is snippy and mean about having to move the item from common loot to uncommon loot

File: code/modules/paperwork/pen.dm [hunk 10]
```diff
@@ -339,3 +339,52 @@
...
+		AddElement(/datum/element/eyestab)
+	update_appearance(UPDATE_ICON)
+	
+/obj/item/pen/screwdriver/attack_self(mob/living/user)
+	toggle_screwdriver(user)
+
+
+/obj/item/pen/screwdriver/update_icon_state()
+	. = ..()
+	icon_state = "[initial(icon_state)][extended ? "out":null]"
```
Comment (ShizCalev) [inline code/modules/paperwork/pen.dm:390]:
```suggestion
	icon_state = "[initial(icon_state)][extended ? "_out":null]"
```

File: code/modules/paperwork/pen.dm [hunk 11]
```diff
@@ -339,3 +339,52 @@
...
 	demolition_mod = 9000
+
+// screwdriver pen!
+
+/obj/item/pen/screwdriver
+	/// whether the pen is extended
+	var/extended = FALSE
+	desc = "A pen with an extendable screwdriver tip. This one has a yellow cap."
+	icon_state = "pendriver"
+	toolspeed = 1.20  // gotta have some downside
```
Comment (Fikou) [inline code/modules/paperwork/pen.dm:350]:
```suggestion
	toolspeed = 1.2  // gotta have some downside
```

File: code/modules/paperwork/pen.dm [hunk 12]
```diff
@@ -339,3 +339,52 @@
 	force = 5
 	wound_bonus = 100
 	demolition_mod = 9000
+
+// screwdriver pen!
+
+/obj/item/pen/screwdriver
```
Comment (Fikou) [inline code/modules/paperwork/pen.dm:345]: make it use the transforming component

Comment (etherware-novice) [reply code/modules/paperwork/pen.dm:345]: the..what?

Comment (Fikou) [reply code/modules/paperwork/pen.dm:345]: see the edagger

File: code/modules/paperwork/pen.dm [hunk 13]
```diff
@@ -349,41 +349,38 @@
 	icon_state = "pendriver"
 	toolspeed = 1.2  // gotta have some downside
 
+/obj/item/pen/screwdriver/Initialize(mapload)
+	. = ..()
+	AddComponent(/datum/component/transforming, \
```
Comment (optimumtact) [inline code/modules/paperwork/pen.dm:354]: I think the \ is breaking this

Comment (optimumtact) [reply code/modules/paperwork/pen.dm:354]: oh no, nevermind, you're just missing a , after WEIGHT_CLASS_SMALL

File: code/modules/modular_computers/computers/item/tablet.dm [hunk 14]
```diff
@@ -54,6 +54,7 @@
 
 	if(is_type_in_list(W, contained_item))
 		if(W.w_class >= WEIGHT_CLASS_SMALL) // Anything equal to or larger than small won't work
+			to_chat(user, span_warning("\The [W] is too big!"))
```
Comment (SuperNovaa41) [inline code/modules/modular_computers/computers/item/tablet.dm:57]: would be nice if this were a balloon alert

**Review by SuperNovaa41 — CHANGES_REQUESTED** [review]

File: code/modules/paperwork/pen.dm [hunk 15]
```diff
@@ -339,3 +339,49 @@
...
+/obj/item/pen/screwdriver/vv_edit_var(var_name, var_value)
+	if(var_name == NAMEOF(src, extended))
+		if(var_value != extended)
+			SEND_SIGNAL(src, COMSIG_ITEM_ATTACK_SELF)
+			datum_flags |= DF_VAR_EDITED
+			return
+	return ..()
+
+/obj/item/pen/screwdriver/proc/toggle_screwdriver(mob/user, active)
+	extended = active
```
Comment (ShizCalev) [inline code/modules/paperwork/pen.dm:371]:
```suggestion
	SIGNAL_HANDLER
	extended = active
```

File: code/modules/paperwork/pen.dm [hunk 16]
```diff
@@ -339,3 +339,50 @@
...
+	else
+		tool_behaviour = TOOL_SCREWDRIVER
+		AddElement(/datum/element/eyestab)
+
+	update_appearance(UPDATE_ICON)
+	return COMPONENT_NO_DEFAULT_MESSAGE
+
+/obj/item/pen/screwdriver/update_icon_state()
+	. = ..()
+	icon_state = "[initial(icon_state)][extended ? "_out":null]"
```
Comment (ShizCalev) [inline code/modules/paperwork/pen.dm:388]:
```suggestion
	icon_state = "[initial(icon_state)][extended ? "_out":null]"
	inhand_icon_state = [initial(inhand_icon_state)]
```

File: code/modules/paperwork/pen.dm [hunk 17]
```diff
@@ -339,3 +339,50 @@
...
+	var/extended = FALSE
+
+/obj/item/pen/screwdriver/Initialize(mapload)
+	. = ..()
+	AddComponent(/datum/component/transforming, \
+		throwforce_on = 5, \
+		w_class_on = WEIGHT_CLASS_SMALL, \
+		sharpness_on = TRUE)
+
+	RegisterSignal(src, COMSIG_TRANSFORMING_ON_TRANSFORM, .proc/toggle_screwdriver)
```
Comment (ShizCalev) [inline code/modules/paperwork/pen.dm:359]:
```suggestion
	RegisterSignal(src, COMSIG_TRANSFORMING_ON_TRANSFORM, .proc/toggle_screwdriver)
	AddElement(/datum/element/update_icon_updates_onmob)
```

**Review by Imaginos16 — CHANGES_REQUESTED** [review]
Please change the pen sprite to follow our standards. They currently do not match the style of all other pens in game right now.

**Comment by etherware-novice** [comment]
what does the update_icon_update_onmob component do?

**Comment by ShizCalev** [comment]
it updates the held icons since the transforming component overrides them.

File: code/modules/paperwork/pen.dm [hunk 18]
```diff
@@ -339,3 +339,52 @@
...
+	AddElement(/datum/element/update_icon_updates_onmob)
+
+
+/obj/item/pen/screwdriver/vv_edit_var(var_name, var_value)
+	if(var_name == NAMEOF(src, extended))
+		if(var_value != extended)
+			SEND_SIGNAL(src, COMSIG_ITEM_ATTACK_SELF)
```
Comment (Mothblocks) [inline code/modules/paperwork/pen.dm:366]: This is weird. Why are you doing this?

Comment (Mothblocks) [reply code/modules/paperwork/pen.dm:366]: Can this not just call `attack_self()` normally?

**Review by Mothblocks — CHANGES_REQUESTED** [review]
@Imaginos16 Is the sprite ok?

File: code/modules/paperwork/pen.dm (hunk 18 above)
Comment (etherware-novice) [reply code/modules/paperwork/pen.dm:366]: does calling attack self trigger the transform component?

Comment (Mothblocks) [reply code/modules/paperwork/pen.dm:366]: I don't know, but doing this is definitely wrong.

File: code/modules/paperwork/pen.dm [hunk 19]
```diff
@@ -339,3 +339,52 @@
...
+		sharpness_on = TRUE)
+
+	RegisterSignal(src, COMSIG_TRANSFORMING_ON_TRANSFORM, .proc/toggle_screwdriver)
+	AddElement(/datum/element/update_icon_updates_onmob)
+
+
+/obj/item/pen/screwdriver/vv_edit_var(var_name, var_value)
+	if(var_name == NAMEOF(src, extended))
+		if(var_value != extended)
+			attack_self()
```
Comment (Fikou) [inline code/modules/paperwork/pen.dm:366]: toggle_screwdriver instead?

Comment (etherware-novice) [reply code/modules/paperwork/pen.dm:366]: then the transform component is completely skipped, and toggle_screwdriver does nothing besides toggle a few vars that arent managed by transform

Comment (ShizCalev) [reply code/modules/paperwork/pen.dm:366]: sec

**Review by Imaginos16 — CHANGES_REQUESTED** [review]
Its still not to my liking. If you would like to actually discuss with me what I want to see changed, contact Imaginos#6760 on discord.

**Review by ShizCalev — APPROVED** [review]

**Comment by etherware-novice** [comment]
i thought using getcomponent was a last resort according to the hackmd page

**Comment by ShizCalev** [comment]
Code is ready, good to merge once sprite is done.

**Comment by LemonInTheDark** [comment]
> i thought using getcomponent was a last resort according to the hackmd page

it is, but sending a signal like that is "worse"
really we should have a vv signal or something but ahhh

**Comment by etherware-novice** [comment]
so.. never directly call attack_self without a user?

**Review by Imaginos16 — APPROVED** [review]
Much better.

**Comment by LemonInTheDark** [comment]
> so.. never directly call attack_self without a user?

I mean yes but also don't do that

**Pull request merged**